### PR TITLE
Refactor image data binding

### PR DESCRIPTION
**Pull request by novain1102**
Re: #https://github.com/cocos/cocos-engine/issues/11235

### Changelog

* 

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.

  > Manual trigger with `@cocos-robot run test cases` afterward.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: native/tools/tojs/cocos.ini [hunk 1]
```diff
@@ -61,7 +61,8 @@ rename_functions = FileUtils::[loadFilenameLookupDictionaryFromFile=loadFilename
                    CCObject::[destroyImmediate=_destroyImmediate destroy=_destroy]
 
 getter_setter=ICanvasRenderingContext2D::[width//setWidth height//setHeight fillStyle//setFillStyle font//setFont globalCompositeOperation//setGlobalCompositeOperation lineCap//setLineCap lineJoin//setLineJoin lineWidth//setLineWidth strokeStyle//setStrokeStyle textAlign//setTextAlign textBaseline//setTextBaseline],
-        CCObject::[name hideFlags replicated/isReplicated/setReplicated isValid/isValid]
+        CCObject::[name hideFlags replicated/isReplicated/setReplicated isValid/isValid],
+        JSBNativeDataHolder::[data]
```
Comment (dumganhar) [inline native/tools/tojs/cocos.ini]: Should bind `data` property?
Should `getData` & `setData` be skipped?

File: native/cocos/bindings/manual/jsb_global.cpp [hunk 2]
```diff
@@ -583,8 +585,9 @@ bool jsb_global_load_image(const ccstd::string &path, const se::Value &callbackV
 
                 if (loadSucceed) {
                     se::HandleObject retObj(se::Object::createPlainObject());
-                    dataVal.setUint64(reinterpret_cast<uintptr_t>(imgInfo->data));
-                    retObj->setProperty("data", dataVal);
+                    auto *obj = se::Object::createObjectWithClass(__jsb_cc_JSBNativeDataHolder_class);
+                    obj->setPrivateData(ccnew cc::JSBNativeDataHolder(imgInfo->data));
```
Comment (dumganhar) [inline native/cocos/bindings/manual/jsb_global.cpp]:
```c++
obj->setPrivateData(ccnew cc::JSBNativeDataHolder(imgInfo->data));
```
needs to changed to 
```c++
auto *nativeObj = JSB_MAKE_PRIVATE_OBJECT(cc::JSBNativeDataHolder, imgInfo->data);
obj->setPrivateObject(nativeObj);
```

File: native/tools/tojs/cocos.ini (hunk 1 above)
Comment (novain1102) [reply native/tools/tojs/cocos.ini]: Removed it.

File: native/cocos/bindings/manual/jsb_global.cpp (hunk 2 above)
Comment (novain1102) [reply native/cocos/bindings/manual/jsb_global.cpp]: what's the difference?

Comment (mmyduckx) [reply native/cocos/bindings/manual/jsb_global.cpp:589]: cannot get `se::state s` paramater in this funciton

**Review by dumganhar — APPROVED** [review]

**Review by PatriceJiang — APPROVED** [review]

**Comment by suqionglian** [comment]
@cocos-robot run test cases

File: native/cocos/bindings/manual/jsb_global.cpp [hunk 3]
```diff
@@ -650,22 +654,23 @@ static bool js_loadImage(se::State &s) { //NOLINT
...
     const auto &args = s.args();
     size_t argc = args.size();
     CC_UNUSED bool ok = true;
     if (argc == 1) {
-        auto *data = reinterpret_cast<char *>(args[0].asPtr());
+        cc::JSBNativeDataHolder *dataHolder = nullptr;
+        ok &= sevalue_to_native(args[0], &dataHolder);
         SE_PRECONDITION2(ok, false, "js_destroyImage : Error processing arguments");
-        free(data);
+//        dataHolder->destroy();
```
Comment (minggo) [inline native/cocos/bindings/manual/jsb_global.cpp]: dataHolde seems useless.

Comment (novain1102) [reply native/cocos/bindings/manual/jsb_global.cpp]: Yes. just to replace old logic by explicit call destroy although resouces will be released by destructor

File: native/cocos/bindings/manual/jsb_global.cpp [hunk 4]
```diff
@@ -583,8 +585,10 @@ bool jsb_global_load_image(const ccstd::string &path, const se::Value &callbackV
 
                 if (loadSucceed) {
                     se::HandleObject retObj(se::Object::createPlainObject());
-                    dataVal.setUint64(reinterpret_cast<uintptr_t>(imgInfo->data));
-                    retObj->setProperty("data", dataVal);
```
Comment (minggo) [inline native/cocos/bindings/manual/jsb_global.cpp:587]: dataVal us not used, then can be deleted?

File: native/cocos/bindings/manual/jsb_global.cpp [hunk 5]
```diff
@@ -650,22 +654,23 @@ static bool js_loadImage(se::State &s) { //NOLINT
 }
 SE_BIND_FUNC(js_loadImage)
 
-static bool js_destroyImage(se::State &s) { //NOLINT
+static bool js_destroyImage(se::State &s) { // NOLINT
```
Comment (minggo) [inline native/cocos/bindings/manual/jsb_global.cpp:656]: Is this function needed any more?

File: native/cocos/bindings/manual/jsb_global.cpp (hunk 3 above)
Comment (dumganhar) [reply native/cocos/bindings/manual/jsb_global.cpp]: Why comment dataHolder->destroy(); ? I think we need it.
`free` in destructor of JSBNativeHolderData is the last step of checking data.
That means:
- If user invokes image.destroy(), then ~JSBNativeHolderData does nothing. And image data could be released in time before GC comes.
- If user forgets to invokes image.destroy(), then ~JSBNativeHolderData free data to make sure there is not leaks.

Comment (minggo) [reply native/cocos/bindings/manual/jsb_global.cpp]: But dataHolder is a pointer, it is defined, but is not used. Then need it any more?

Comment (mmyduckx) [reply native/cocos/bindings/manual/jsb_global.cpp:665]: >  you're right. I'll add it back

Comment (novain1102) [reply native/cocos/bindings/manual/jsb_global.cpp]: > Why comment dataHolder->destroy(); ? I think we need it. `free` in destructor of JSBNativeHolderData is the last step of checking data. That means:
> 
> * If user invokes image.destroy(), then ~JSBNativeHolderData does nothing. And image data could be released in time before GC comes.
> * If user forgets to invokes image.destroy(), then ~JSBNativeHolderData free data to make sure there is not leaks.

you're right. I'll add it back

File: native/cocos/bindings/manual/jsb_global.cpp (hunk 4 above)
Comment (novain1102) [reply native/cocos/bindings/manual/jsb_global.cpp:587]: Yes. removed it

File: native/cocos/bindings/manual/jsb_global.cpp (hunk 5 above)
Comment (novain1102) [reply native/cocos/bindings/manual/jsb_global.cpp:656]: add back destroy

**Review by minggo — APPROVED** [review]

**Pull request merged**